lib/Sema/CSDiag.cpp

```diff
@@ -5421,8 +5431,96 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
     // Type check the function subexpression to resolve a type for it if
     // possible.
     fnExpr = typeCheckChildIndependently(callExpr->getFn());
-    if (!fnExpr)
-      return true;
+    if (!fnExpr) {
+      auto hadError = [&]() {
+        return CS.TC.Diags.hadAnyError();
+      };
+      auto UDE = dyn_cast<UnresolvedDotExpr>(callExpr->getFn());
+      if (!UDE)
+        return hadError();
+
+      auto baseExpr = UDE->getBase();
+      // Offer fix-it for using `subscript` instead of the operator.
+      // FIXME: This doesn't check generic parameters
+      if (baseExpr && UDE->getName().getBaseName() == "subscript") {
+        auto baseType = CS.getType(baseExpr);
+        // Look up subscript declarations and return if there are none.
+        auto lookup = CS.lookupMember(baseType->getRValueType(),
+                                      DeclName(DeclBaseName::createSubscript()));
+        if (lookup.empty())
+          return hadError();
+
+        auto argExpr = typeCheckChildIndependently(callExpr->getArg(), Type(),
+                                                   CTP_CallArgument);
+        if (!argExpr)
+          return hadError();
+
+        SmallVector<Identifier, 2> scratch;
+        ArrayRef<Identifier> argLabels = callExpr->getArgumentLabels(scratch);
+        SmallVector<OverloadChoice, 2> choices;
+        auto hasTrailClosure = callArgHasTrailingClosure(argExpr);
+
+        for (auto candidate : lookup)
+          choices.push_back(OverloadChoice(baseType, candidate.getValueDecl(),
+                                           UDE->getFunctionRefKind()));
+
+        CalleeCandidateInfo candidates(baseType, choices,
+                                       hasTrailClosure, CS, true);
+
+        auto argType = CS.getType(argExpr);
+        auto params = swift::decomposeArgType(argType, argLabels);
+
+        using Closeness = CalleeCandidateInfo::ClosenessResultTy;
+
+        candidates.filterList([&](UncurriedCandidate candidate) -> Closeness {
+          auto candFuncType = candidate.getUncurriedFunctionType();
+          if (!candFuncType)
+            return {CC_GeneralMismatch, {}};
+
+          auto candParams = candFuncType->getParams();
+
+          if (params.size() != candParams.size())
+            return {CC_GeneralMismatch, {}};
+
+          for (unsigned i = 0, e = params.size(); i < e; i ++) {
+            auto type1 = params[i].getType()->
+              getUnlabeledType(CS.getASTContext());
+            auto type2 = candParams[i].getType()->
+              getUnlabeledType(CS.getASTContext());
+
+            if (type1->isEqual(type2) || type2->is<GenericTypeParamType>() ||
```

Comment:
Is `type2->is<GenericTypeParamType>()` enough?  I think we want to check if `type1` can be bound to `type2`.

// @xedin

Reply:
I think @AnthonyLatsis mentioned that it doesn't really work for generic parameters as of now, which I think is okay.

Reply:
@CodaFi @xedin to be bound here means to be substitutable or bound a in T=U?

Reply:
If I write the logic for generics as well you can almost consider this a new `evaluateCloseness` :)